File: client/src/app/models/speaker.ts
import { FormControl, FormGroup, Validators } from '@angular/forms';

export class Speaker {
    id: String;
    name: String;
    contact: String;
    title: String;
    bio: String;
    notes: String;
    imgs: SpeakerImages;
    participations: SpeakerParticipation[];
}

export class SpeakerImages {
    internal: String;
    speaker: String;
    company: String;
}

export class SpeakerParticipation {
    event: Number;
    member: String;
    status: String;
    communications: String[];
    subscribers: String[];
    feedback: String;
    flights: String[];
    room: SpeakerParticipationRoom;
}

export class SpeakerParticipationRoom {
    type: String;
    cost: Number;
    notes: String;
}

export class SpeakerParticipationValidStatusSteps {
    steps: {
        next: String;
        step: Number;
    }[];
}

export function GetParticipation(speaker: Speaker, event: Number): SpeakerParticipation {
    for (const participation of speaker.participations) {
        if (participation.event === event) {
            return participation;
        }
    }

    return null;
}

export class EditSpeakerForm {

    form: FormGroup;

    constructor(speaker: Speaker) {
        this.form = new FormGroup({
            bio: new FormControl(speaker.bio ? speaker.bio : '', [Validators.required, Validators.minLength(1)]),
            name: new FormControl(speaker.name, [Validators.required, Validators.minLength(1)]),
            notes: new FormControl(speaker.notes, [Validators.required, Validators.minLength(1)]),
            title: new FormControl(speaker.title, [Validators.required, Validators.minLength(1)]),
        });
    }

    value() {
        return this.form.value;
    }

    valid() {
        return this.form.valid;
    }

}
export class AddSpeakerForm {

    form: FormGroup;

    constructor() {
        this.form = new FormGroup({
            name: new FormControl('', [Validators.required, Validators.minLength(1)]),
            bio: new FormControl('', []),
            title: new FormControl('', [Validators.required, Validators.minLength(1)]),
        });
    }

    value() {
        return this.form.value;
    }

    valid() {
        return this.form.valid;
    }
}

export class EditSpeakerParticipationForm {

    form: FormGroup;

    constructor(participation?: SpeakerParticipation) {
        this.form = new FormGroup({
            feedback: new FormControl(participation ? participation.feedback : '', []),
            member: new FormControl(participation ? participation.member : '', [Validators.required, Validators.minLength(1)]),
            room: new FormGroup({
                cost: new FormControl(participation && participation.room ? participation.room.cost : 0, [Validators.required, Validators.min(0)]),
                notes: new FormControl(participation && participation.room ? participation.room.notes : '', []),
                type: new FormControl(participation && participation.room ? participation.room.type : '', []),
            })
        });
    }

    value() {
        return this.form.value;
    }

    valid() {
        return this.form.valid;
    }
}

export class EditSpeakerImageForm {

    form: FormGroup;
    file: File;

    constructor() {
        this.form = new FormGroup({
            image: new FormControl(null, [Validators.required]),
        });
    }

    value() {
        return this.file;
    }

    valid() {
        return this.form.valid;
    }

    set(file: File) {
        this.file = file;
    }
}

export class EditSpeakerParticipationStatusForm {

    form: FormGroup;

    constructor() {
        this.form = new FormGroup({});
    }

    valid() {
        return this.form.valid;
    }
}

File: backend/src/models/speaker.go
package models

import "go.mongodb.org/mongo-driver/bson/primitive"

type SpeakerParticipationRoom struct {

	// Small description of the booking.
	Type string `json:"type" bson:"type"`

	// Cost of the booking in cents (€).
	Cost int `json:"cost" bson:"cost"`

	// Additional information, like which hotel, its address, etc.
	Notes string `json:"notes" bson:"notes"`
}

type SpeakerParticipation struct {

	// Event is an _id of Event (see models.Event).
	Event int `json:"event" bson:"event"`

	// Member is an _id of Member (see models.Member).
	Member primitive.ObjectID `json:"member" bson:"member"`

	// Participation's status. See models.Company
	Status ParticipationStatus `json:"status" bson:"status"`

	// Communications is an array of _id of Communication (see models.Communication).
	Communications []primitive.ObjectID `json:"communications" bson:"communications"`

	// Subscribers is an array of _id of Member (see models.Member).
	Subscribers []primitive.ObjectID `json:"subscribers" bson:"subscribers"`

	// Feedback given by the speaker regarding the conference. This will be visible on
	// our website.
	Feedback string `json:"feedback" bson:"feedback"`

	// Flights is an array of _id of FlightInfo (see models.FlightInfo).
	Flights []primitive.ObjectID `json:"flights" bson:"flights"`

	// Hotel information regarding this speaker.
	Room SpeakerParticipationRoom `json:"room" bson:"room"`
}

type SpeakerImages struct {

	// Internal image, only visible by the team.
	Internal string `json:"internal" bson:"internal"`

	// Speaker photo. This is a URL pointing to the photo, stored somewhere. It will
	// be visible on our website.
	Speaker string `json:"speaker" bson:"speaker"`

	// Speaker's company photo. This is a URL pointing to the photo, stored somewhere. It will
	// be visible on our website.
	Company string `json:"company" bson:"company"`
}

type Speaker struct {

	// Speaker's ID (_id of mongodb).
	ID primitive.ObjectID `json:"id" bson:"_id"`

	Name string `json:"name" bson:"name"`

	// Contact is an _id of Contact (see models.Contact).
	Contact *primitive.ObjectID `json:"contact,omitempty" bson:"contact"`

	// Title of the speaker (CEO @ HugeCorportation, for example).
	Title string `json:"title" bson:"title"`

	// Bio of the speaker. Careful, this will be visible on our website!
	Bio string `json:"bio" bson:"bio"`

  // Company name
  CompanyName string `json:"companyName" bson:"companyName"`

	// This is only visible by the team. Praise and trash talk at will.
	Notes          string                 `json:"notes" bson:"notes"`
	Images         SpeakerImages          `json:"imgs" bson:"imgs"`
	Participations []SpeakerParticipation `json:"participations" bson:"participations"`
}

type SpeakerImagesPublic struct {

	// Speaker photo. This is a URL pointing to the photo, stored somewhere. It will
	// be visible on our website.
	Speaker string `json:"speaker" bson:"speaker"`

	// Speaker's company photo. This is a URL pointing to the photo, stored somewhere. It will
	// be visible on our website.
	Company string `json:"company,omitempty" bson:"company"`
}

type SpeakerParticipationPublic struct {

	// Event is an _id of Event (see models.Event).
	Event int `json:"event" bson:"event"`

	// Feedback given by the speaker regarding the conference. This will be visible on
	// our website.
	Feedback string `json:"feedback" bson:"feedback"`
}

// SpeakerPublic represents a speaker to be contacted by the team, that will hopefully participate
// on the event, and to be shown to everyone publicly.
type SpeakerPublic struct {
	ID primitive.ObjectID `json:"id" bson:"_id"`

	Name string `json:"name" bson:"name"`

	// Title of the speaker (CEO @ HugeCorportation, for example).
	Title string `json:"title" bson:"title"`

	// Bio of the speaker. Careful, this will be visible on our website!
	Bio string `json:"bio" bson:"bio"`

  // Company name
  CompanyName string `json:"companyName,omitempty" bson:"companyName"`

	Images         SpeakerImagesPublic          `json:"imgs" bson:"imgs"`
	Participations []SpeakerParticipationPublic `json:"participation" bson:"participations"`
}
